File: go/fixture/go-mini/internal/pool/workers.go
// Package pool runs jobs on a fixed set of worker goroutines.
package pool

import (
	"log"

	"example.com/go-mini/internal/env"
)

// Job is a unit of work handed to a worker.
type Job struct {
	ID  string
	Run func()
}

// Start starts the workers.
func Start(jobs <-chan Job) {
	for i := 0; i < env.Config.NumWorkers; i++ {
		go worker(jobs)
	}
}

func worker(jobs <-chan Job) {
	for j := range jobs {
		if j.Run != nil {
			j.Run()
		}
		log.Printf("pool: job %s done", j.ID)
	}
}

File: py/fixture/py-mini/internal/pool/workers.py
"""Runs jobs on a fixed set of worker threads."""

import logging
import queue
import threading
from collections.abc import Callable
from dataclasses import dataclass

from internal.env import env

_LOG = logging.getLogger("pool")


@dataclass
class Job:
    """A unit of work handed to a worker."""

    id: str
    run: Callable[[], None] | None = None


def start(jobs: queue.Queue[Job | None]) -> int:
    """Start the workers and return how many were started."""
    for _ in range(env.CONFIG.num_workers):
        threading.Thread(target=_worker, args=(jobs,), daemon=True).start()
    return env.CONFIG.num_workers


def _worker(jobs: queue.Queue[Job | None]) -> None:
    for job in iter(jobs.get, None):
        if job.run is not None:
            job.run()
        _LOG.info("pool: job %s done", job.id)
